providers/discord.go

```diff
@@ -0,0 +1,111 @@
+package providers
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/bitly/oauth2_proxy/api"
+)
+
+type DiscordProvider struct {
+	*ProviderData
+}
+
+type DiscordUserInfo struct {
+	Id            string `json:"id"`
+	Avatar        string `json:"avatar"`
+	Username      string `json:"username"`
+	Discriminator string `json:"discriminator"`
+	Verified      bool   `json:"verified"`
+	Email         string `json:"email"`
+	Flags         int    `json:"flags"`
+	Locale        string `json:"locale"`
+	PremiumType   int    `json:"premium_type"`
+	MfaEnabled    bool   `json:"mfa_enabled"`
+}
+
+func NewDiscordProvider(p *ProviderData) *DiscordProvider {
+	p.ProviderName = "Discord"
+	if p.LoginURL.String() == "" {
+		p.LoginURL = &url.URL{Scheme: "https",
+			Host: "discordapp.com",
+			Path: "/api/oauth2/authorize",
+		}
+	}
+	if p.RedeemURL.String() == "" {
+		p.RedeemURL = &url.URL{Scheme: "https",
+			Host: "discordapp.com",
+			Path: "/api/oauth2/token",
+		}
+	}
+	if p.ProfileURL.String() == "" {
+		p.ProfileURL = &url.URL{Scheme: "https",
+			Host: "discordapp.com",
+			Path: "/api/users/@me",
+		}
+	}
+	if p.ValidateURL.String() == "" {
+		p.ValidateURL = p.ProfileURL
+	}
+	if p.Scope == "" {
+		p.Scope = "identify email connections"
```

Comment:
looks like "connections" is not needed

Reply:
this and some other little refinements could be in a follow-up pull request if someone cares enough :)

(e.g. removing parts of DiscordUserInfo which are unused, maybe checking Verified property ...)